File: baekjoon_all/01000+/boj_1927_solution2.cpp
// Solve 2022-07-27
// Update 2024-02-11

#include <bits/stdc++.h>
using namespace std;

#define FASTIO ios_base::sync_with_stdio(false);cin.tie(NULL); // boj_15552.cpp
#define SETPRECISION(n) cout << fixed;cout.precision(n); // boj_1008.cpp
#define SIZE(v) (int)v.size()
#define ALL(v) v.begin(),v.end()
using ll = long long;

int heap[100001];
int heap_end = 1;

void push(int x) {
    int cur_node = heap_end++;
    int cur_parent = cur_node / 2;

    while (cur_parent >= 1 && heap[cur_parent] > x) {
        heap[cur_node] = heap[cur_parent];
        cur_node = cur_parent;
        cur_parent /= 2;
    }

    heap[cur_node] = x;
}

int pop() {
    if (heap_end == 1) return 0;

    int value_return = heap[1];
    int value_move = heap[--heap_end];
    int cur_node = 1;
    int cur_child = 2;

    while (cur_child < heap_end) {
        if (cur_child + 1 < heap_end && heap[cur_child] > heap[cur_child + 1]) {
            cur_child++;
        }

        if (value_move <= heap[cur_child]) break;

        heap[cur_node] = heap[cur_child];
        cur_node = cur_child;
        cur_child *= 2;
    }

    heap[cur_node] = value_move;
    return value_return;
}

int main() {
    FASTIO;

    int n;
    cin >> n;

    while (n-- > 0) {
        int x;
        cin >> x;

        if (x == 0) cout << pop() << '\n';
        else push(x);
    }

    return 0;
}

File: baekjoon_all/01000+/boj_1927_solution2.java
// Solve 2024-02-11

// 백준에 제출할 때는 class 이름을 Main으로 설정해야 한다.

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class boj_1927_solution2 {

    static class PqMinTop {
        private int[] heap;
        private int heapEnd;

        PqMinTop(int capacity) {
            heap = new int[capacity + 1];
            heapEnd = 1;
        }

        boolean isEmpty() {
            return heapEnd == 1;
        }

        int size() {
            return heapEnd - 1;
        }

        void push(int x) {
            int curNode = heapEnd++;
            int curParent = curNode / 2;

            while (curParent >= 1 && heap[curParent] > x) {
                heap[curNode] = heap[curParent];
                curNode = curParent;
                curParent /= 2;
            }

            heap[curNode] = x;
        }

        int pop() {
            if (isEmpty()) return 0;

            int valueReturn = heap[1];
            int valueMove = heap[--heapEnd];
            int curNode = 1;
            int curChild = 2;

            while (curChild < heapEnd) {
                if (curChild + 1 < heapEnd && heap[curChild] > heap[curChild + 1]) {
                    curChild++;
                }

                if (valueMove <= heap[curChild]) break;

                heap[curNode] = heap[curChild];
                curNode = curChild;
                curChild *= 2;
            }

            heap[curNode] = valueMove;
            return valueReturn;
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        int n = Integer.parseInt(br.readLine());
        PqMinTop pqMinTop = new PqMinTop(100_000);

        while (n-- > 0) {
            int x = Integer.parseInt(br.readLine());

            if (x == 0) {
                sb.append(pqMinTop.pop()).append("\n");
            } else {
                pqMinTop.push(x);
            }
        }

        System.out.print(sb);
        br.close();
    }

}
